### PR TITLE
text tag

### PR DESCRIPTION
**Pull request by mgrudzinska**
todo:
- default values handling
- font-family multi values
- fix grad?
- meson - better idea?

Zrzut ekranu 2024-05-6 o 00 43 31

```
<svg viewBox="0 0 125 115">
  <defs>
    <linearGradient id="rainbow" x1="0" x2="0" y1="20%" y2="100%" gradientUnits="userSpaceOnUse" >
      <stop stop-color="#ff00ff" offset="0%"/>
      <stop stop-color="#FF5447" offset="30%"/>
      <stop stop-color="#FF7B21" offset="40%"/>
      <stop stop-color="#EAFC37" offset="60%"/>
      <stop stop-color="#4FCB6B" offset="80%"/>
      <stop stop-color="#0000ff" offset="100%"/> 
    </linearGradient>
  </defs>
    <text font-family="Monaco" fill="#ff00ff" font-size="25" x="0" y="30">gradient</text>
    <text font-family="Monaco" fill="url(#rainbow)" font-size="45" x="0" y="70">text</text>
    <text font-family="Monaco" font-size="20" x="30" y="100">in svg</text>
    <rect x="120" y="0" width="5" height="115" fill="url(#rainbow)"/>
</svg>
```

issue: https://github.com/thorvg/thorvg/issues/1897

### CODE REVIEW

File: src/loaders/svg/tvgSvgLoader.cpp [hunk 1]
```diff
@@ -3308,6 +3386,15 @@ static void _svgLoaderParserXmlOpen(SvgLoaderData* loader, const char* content,
 }
 
 
+static void _svgLoaderParserText(SvgLoaderData* loader, const char* content, unsigned int length)
+{
+    auto text = &loader->svgParse->node->node.text;
+    if (text->text) free(text->text);
+    text->text = strndup(content, length);
```
Comment (JSUYA) [inline src/loaders/svg/tvgSvgLoader.cpp:3393]: I recommend to use strDuplicate().

File: inc/thorvg.h [hunk 2]
```diff
@@ -1485,7 +1486,7 @@ class TVG_API Text final : public Paint
      *
      * @note Experimental API
      */
-    Result fill(uint8_t r, uint8_t g, uint8_t b) noexcept;
+    Result fill(uint8_t r, uint8_t g, uint8_t b, uint8_t a = 255) noexcept;
```
Comment (hermet) [inline inc/thorvg.h:1489]: it has opacity() already ?

**Comment by hermet** [comment]
@mgrudzinska ping?

**Comment by mgrudzinska** [comment]
@hermet I still have to figure out how to access the system fonts or how to deal without them

**Comment by hermet** [comment]
@mgrudzinska Please handle the embedded font as expected for the issue. System fonts must be addressed separately based on the thorvg specifications.

**Comment by mgrudzinska** [comment]
@hermet I keep it as draft, since it contains [#2345](https://github.com/thorvg/thorvg/pull/2345) and it's not clear for now whether we want it, so there's no point to push this one into review

File: src/loaders/svg/tvgSvgSceneBuilder.cpp [hunk 3]
```diff
@@ -771,6 +814,32 @@ static unique_ptr<Scene> _useBuildHelper(SvgLoaderData& loaderData, const SvgNod
...
+        Matrix translate = {1, 0, node->node.text.x, 0, 1, node->node.text.y - textNode->fontSize, 0, 0, 1};
+        textTransform *= translate;
+    }
+    text->transform(textTransform);
+
+    const float ptPerPx = 0.75f; //1 pt = 1/72; 1 in = 96 px; -> 72/96 = 0.75
+    auto fontSizePt = textNode->fontSize * ptPerPx;
+    //TODO: handle default fonts
+    if (textNode->fontFamily) text->font(textNode->fontFamily, fontSizePt);
+    text->text(textNode->text);
```
Comment (mgrudzinska) [inline src/loaders/svg/tvgSvgSceneBuilder.cpp:834]: check if ```textNode->text``` if null, if so then exit earlier

**Comment by mgrudzinska** [comment]
in a final version it will handle embedded fonts only, the rest in handled in:
https://github.com/thorvg/thorvg/pull/2345
https://github.com/thorvg/thorvg/pull/2362

**Pull request closed without merging**